### PR TITLE
add yaml schema validator

### PR DESCRIPTION
**Pull request by MohamedFarouk-HMCTS**

### CODE REVIEW

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidator.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    public SubscriptionDescriptorFileValidator(final YamlFileToJsonObjectConverter yamlFileToJsonObjectConverter) {
+        this.yamlFileToJsonObjectConverter = yamlFileToJsonObjectConverter;
+    }
+
+    public void validate(final Path filePath) {
+        final JSONObject convert;
+        try {
+            convert = yamlFileToJsonObjectConverter.convert(filePath);
+        } catch (IOException ex) {
+            throw new SubscriptionDescriptorException(format("Unable to convert to JSON", filePath.toString()), ex);
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidator.java]: fixed

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParser.java [hunk 2]
```diff
@@ -16,27 +17,40 @@
...
 
-        try {
-            return mapper.readValue(new File(filePath), SubscriptionDescriptorDef.class).getSubscriptionDescriptor();
-        } catch (IOException e) {
-            throw new SubscriptionDescriptorIOException("Failed to read subscriptions yaml file", e);
+            mapper.setPropertyNamingStrategy(PropertyNamingStrategy.CAMEL_CASE_TO_LOWER_CASE_WITH_UNDERSCORES);
+
+            return mapper.readValue(filePath.toFile(), SubscriptionDescriptorDef.class).getSubscriptionDescriptor();
+        } catch (final NoSuchFileException e) {
+            throw new SubscriptionDescriptorException(format("No such subscriptions YAML file", filePath), e);
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParser.java]: fixed

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParserFactory.java [hunk 3]
```diff
@@ -7,6 +7,14 @@
 
     @Override
     public SubscriptionDescriptorFileParser create() {
-        return new SubscriptionDescriptorFileParser();
+        return new SubscriptionDescriptorFileParser(subscriptionDescriptorFileValidator());
+    }
+
+    private YamlFileToJsonObjectConverter yamlFileToJsonObjectConverter() {
```
Comment (matt-rich) [inline subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParserFactory.java]: is there any benefit of having this in a separate method?

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidator.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    public void validate(final Path filePath) {
+        final JSONObject convert;
+        try {
+            convert = yamlFileToJsonObjectConverter.convert(filePath);
+        } catch (IOException ex) {
+            throw new SubscriptionDescriptorException(format("Unable to convert to JSON file %s ", filePath.toString()), ex);
+        }
+        schema().validate(convert);
```
Comment (matt-rich) [inline subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidator.java]: can't this and the declaration of convert all just be in the try block?

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidator.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    private Schema schema() {
+        final String schemaFile = "/json/schema/subscription-schema.json";
```
Comment (matt-rich) [inline subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidator.java]: should this be a constant?

File: subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParserTest.java [hunk 6]
```diff
@@ -14,38 +14,50 @@
...
-    public void shouldFailOnIncorrectSubscriptionYaml() throws URISyntaxException {
+    public void shouldFailOnIncorrectSubscriptionYaml() {
+        final SubscriptionDescriptorFileParser subscriptionDescriptorFileParser = new SubscriptionDescriptorFileParserFactory().create();
 
         final List<Path> paths = asList(get("incorrect-subscription.yaml"));
         try {
-            final Collection<SubscriptionDescriptor> subscriptionDescriptors = parser.parse(baseDir, paths);
+            subscriptionDescriptorFileParser.parse(baseDir, paths);
             fail();
         } catch (RuntimeException re) {
```
Comment (mapingo) [inline subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParserTest.java]: The test should be catching the expected exception and asserting that it is the exception thrown

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/YamlFileToJsonObjectConverter.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+    public JSONObject convert(final Path filePath) throws IOException {
+
+        final Object yamlObject = new ObjectMapper(new YAMLFactory())
+                .readValue(yamlString(filePath), Object.class);
+        final ObjectMapper jsonWriter = new ObjectMapper();
+        return new JSONObject(jsonWriter.writeValueAsString(yamlObject));
+    }
+
+    private String yamlString(Path filePath) throws IOException {
```
Comment (matt-rich) [inline subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/YamlFileToJsonObjectConverter.java]: final

File: subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParserTest.java [hunk 8]
```diff
@@ -14,38 +14,50 @@
...
-    final SubscriptionDescriptorFileParser parser = new SubscriptionDescriptorFileParser();
 
+    @Test
+    public void shouldThrowFileNotFoundException() {
+        final SubscriptionDescriptorFileParser subscriptionDescriptorFileParser = new SubscriptionDescriptorFileParserFactory().create();
+        final List<Path> paths = asList(get("no-subscription.yaml"));
+        try {
+            subscriptionDescriptorFileParser.parse(baseDir, paths);
+            fail();
+        } catch (RuntimeException re) {
```
Comment (mapingo) [inline subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParserTest.java]: The test should be catching the expected exception and asserting that it is the exception thrown

File: subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidatorTest.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.io.IOException;
+
+import org.everit.json.schema.ValidationException;
+import org.junit.Test;
+
+public class SubscriptionDescriptorFileValidatorTest {
+
+    @Test
+    public void shouldPassOnCorrectSubscriptionYaml() throws IOException {
```
Comment (mapingo) [inline subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidatorTest.java]: Maybe be better to say should not throw exception for valid subscription Yaml

File: subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidatorTest.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        try {
+            final SubscriptionDescriptorFileValidator subscriptionDescriptorFileValidator
+                    = new SubscriptionDescriptorFileValidator(new YamlFileToJsonObjectConverter());
+
+            subscriptionDescriptorFileValidator.validate(get("src/test/resources/subscription.yaml"));
+        } catch (ValidationException e) {
+            fail("Validation Failure");
```
Comment (mapingo) [inline subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidatorTest.java]: Maybe a better description would be a good idea here

File: subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/YamlFileToJsonObjectConverterTest.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @Test
+    public void shouldFailToConvertYamlToJsonObjectAUnavailableFile() {
+        final YamlFileToJsonObjectConverter converter = new YamlFileToJsonObjectConverter();
+        try {
+            converter.convert(get("src/test/resources/no-subscription.yaml"));
+            fail("Failure, Converted a unavailable file to JSON Object");
+        } catch (IOException e) {
+            MatcherAssert.assertThat(e, is(instanceOf(IOException.class)));
```
Comment (mapingo) [inline subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/YamlFileToJsonObjectConverterTest.java]: MatcherAssert can be a static import

**Review by matt-rich — CHANGES_REQUESTED** [review]
So we are converting the subscription.yaml into a json object then checking it against the predefined subscription json schema?

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParserFactory.java (hunk 3 above)
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParserFactory.java]: Inlined

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidator.java (hunk 4 above)
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidator.java]: Agreed

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidator.java (hunk 5 above)
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidator.java]: Yes changed

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/YamlFileToJsonObjectConverter.java (hunk 7 above)
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/YamlFileToJsonObjectConverter.java]: removed

File: subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParserTest.java (hunk 6 above)
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParserTest.java]: Fixed

File: subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParserTest.java (hunk 8 above)
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParserTest.java]: Fixed

File: subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidatorTest.java (hunk 9 above)
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidatorTest.java]: fixed

File: subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidatorTest.java (hunk 10 above)
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileValidatorTest.java]: FIxed

File: subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/YamlFileToJsonObjectConverterTest.java (hunk 11 above)
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/test/java/uk/gov/justice/maven/generator/io/files/parser/YamlFileToJsonObjectConverterTest.java]: fixed

**Review by matt-rich — APPROVED** [review]

**Pull request merged**